### PR TITLE
refactor(#2830): multi instance nvim-tree.marks

### LINKED ISSUE #2830: Multi Instance: Refactor: nvim-tree.marks
* #2255 

`nvim-tree.marks` is a stateful singleton. Make it an instantiable class owned by `Explorer`.

All usages must go through the `TreeExplorer` singleton.

### PR DESCRIPTION
**Pull request by alex-courtis**
fixes #2830

### CODE REVIEW

**Comment by alex-courtis** [comment]
Your review would be most gratefully appreciated @raaymax 

I've invited you to the project for easier PRs and issue management.

File: lua/nvim-tree/api.lua [hunk 1]
```diff
@@ -72,10 +73,36 @@ end
 local function wrap_node_or_nil(fn)
   return function(node, ...)
     node = node or lib.get_node_at_cursor()
-    fn(node, ...)
+    return fn(node, ...)
```
Comment (alex-courtis) [inline lua/nvim-tree/api.lua:76]: existing bug

File: lua/nvim-tree/api.lua [hunk 2]
```diff
@@ -72,10 +73,36 @@ end
...
+---@param member_method string method name to invoke on member
+---@return fun(...) : any
+local function wrap_explorer_member(explorer_member, member_method)
+  return wrap(function(...)
+    local explorer = core.get_explorer()
+    if explorer then
+      return explorer[explorer_member][member_method](explorer[explorer_member], ...)
+    end
+  end)
+end
```
Comment (alex-courtis) [inline lua/nvim-tree/api.lua:104]: This isn't the final state, but will allow for easier class creation.

File: lua/nvim-tree/renderer/decorator/bookmarks.lua [hunk 3]
```diff
@@ -34,7 +34,7 @@ end
 ---@param node Node
 ---@return HighlightedString[]|nil icons
 function DecoratorBookmarks:calculate_icons(node)
-  if marks.get_mark(node) then
+  if core.get_explorer() and core.get_explorer().marks:get_mark(node) then
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/decorator/bookmarks.lua:37]: Impractical to access Explorer at this point.

File: lua/nvim-tree/explorer/filters.lua [hunk 4]
```diff
@@ -155,8 +154,11 @@ function M.prepare(git_status)
     status.bufinfo = vim.fn.getbufinfo { buflisted = 1 }
   end
 
-  for _, node in pairs(marks.get_marks()) do
-    status.bookmarks[node.absolute_path] = node.type
+  local explorer = require("nvim-tree.core").get_explorer()
+  if explorer then
+    for _, node in pairs(explorer.marks:get_marks()) do
+      status.bookmarks[node.absolute_path] = node.type
+    end
```
Comment (alex-courtis) [inline lua/nvim-tree/explorer/filters.lua:161]: Impractical to access Explorer at this point.

**Review by raaymax — COMMENTED** [review]
Seen, applied changes to my PR. Seems legit to me.

**Comment by alex-courtis** [comment]
> Seen, applied changes to my PR. Seems legit to me.

Many thanks, I'll let you go first - you're making all the progress.

**Comment by alex-courtis** [comment]
Pushing this one through; it's been soaking for a week and won't interfere with the other 2 WIP.

**Pull request merged**